Mypy disallows overriding an attribute with a property

For example:

```python
class Foo:
    def __init__(self) -> None:
        self.x = 42

class Bar(Foo):
    def __init__(self) -> None:
        self.y = 9000

    @property
    def x(self) -> int:
        print('got x!!!')
        return self.y

    @x.setter
    def x(self, value: int) -> None:
        print('set x!!!')
        self.y = value
```

This should be allowed as it doesn't violate LSP as long as `x` implements both a `getter` and a `setter` of the correct type.

There is a different error message when using `property()` as a function:

```python
class Base:
    prop: bool

class Sub(Base):
    def _get_prop(self) -> bool:
        return False
    def _set_prop(self, value: bool) -> None:
        pass
    prop = property(_get_prop, _set_prop)
```

```
foo.py:9: error: Incompatible types in assignment (expression has type "property", base class "Base" defined the type as "bool")
```

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -1786,9 +1786,13 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                     and isinstance(defn, Decorator)
                 ):
                     # We only give an error where no other similar errors will be given.
-                    self.msg.fail(
-                        "Cannot override writeable attribute with read-only property", defn
-                    )
+                    if not isinstance(original_type, AnyType):
+                        self.msg.fail(
+                            "Cannot override writeable attribute with read-only property",
+                            # Give an error on function line to match old behaviour.
+                            defn.func,
+                            code=codes.OVERRIDE,
+                        )
 
             if isinstance(original_type, AnyType) or isinstance(typ, AnyType):
                 pass

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -1723,12 +1723,10 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
             # TODO: this logic is much less complete than similar one in checkmember.py
             if isinstance(defn, (FuncDef, OverloadedFuncDef)):
                 if isinstance(defn, OverloadedFuncDef) and defn.is_property:
-                    # Type for a settable property is stored in a non-trivial way.
-                    first = defn.items[0]
-                    assert isinstance(first, Decorator)
-                    first_type = first.var.type
-                    assert isinstance(first_type, ProperType) and isinstance(first_type, CallableType)
-                    typ: Type = first_type.ret_type
+                    property_type = get_settable_property_type(defn)
+                    # TODO: are this and below asserts correct, what if decorator was deferred?
+                    assert property_type is not None
+                    typ = property_type
                 else:
                     typ = self.function_type(defn)
                 override_class_or_static = defn.is_class or defn.is_static
@@ -1737,6 +1735,9 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 assert defn.var.is_ready
                 assert defn.var.type is not None
                 typ = defn.var.type
+                if defn.func.is_property:
+                    assert isinstance(typ, ProperType) and isinstance(typ, CallableType)
+                    typ = typ.ret_type
                 override_class_or_static = defn.func.is_class or defn.func.is_static
                 override_class = defn.func.is_class
             typ = get_proper_type(typ)
@@ -1744,8 +1745,15 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 typ = bind_self(typ, self.scope.active_self_type(), is_classmethod=override_class)
             # Map the overridden method type to subtype context so that
             # it can be checked for compatibility.
-            original_type = get_proper_type(base_attr.type)
+            if isinstance(base_attr.node, OverloadedFuncDef) and base_attr.node.is_property:
+                original_type = get_settable_property_type(base_attr.node)
+            else:
+                original_type = base_attr.type
+            original_type = get_proper_type(original_type)
             original_node = base_attr.node
+            if isinstance(original_node, Decorator) and original_node.func.is_property:
+                if isinstance(original_type, CallableType):
+                    original_type = get_proper_type(original_type.ret_type)
             # `original_type` can be partial if (e.g.) it is originally an
             # instance variable from an `__init__` block that becomes deferred.
             if original_type is None or isinstance(original_type, PartialType):
@@ -6981,6 +6989,15 @@ def is_static(func: FuncBase | Decorator) -> bool:
     elif isinstance(func, FuncBase):
         return func.is_static
     assert False, f"Unexpected func type: {type(func)}"
+
+
+def get_settable_property_type(defn: OverloadedFuncDef) -> Type | None:
+    assert isinstance(defn.items[0], Decorator)
+    first_type = defn.items[0].var.type
+    if first_type is None:
+        return None
+    assert isinstance(first_type, ProperType) and isinstance(first_type, CallableType)
+    return first_type.ret_type
 
 
 def is_subtype_no_promote(left: Type, right: Type) -> bool:

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -115,6 +115,7 @@ from mypy.nodes import (
     ReturnStmt,
     StarExpr,
     Statement,
+    SymbolNode,
     SymbolTable,
     SymbolTableNode,
     TempNode,
@@ -1722,22 +1723,13 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
             # Construct the type of the overriding method.
             # TODO: this logic is much less complete than similar one in checkmember.py
             if isinstance(defn, (FuncDef, OverloadedFuncDef)):
-                if isinstance(defn, OverloadedFuncDef) and defn.is_property:
-                    property_type = get_settable_property_type(defn)
-                    # TODO: are this and below asserts correct, what if decorator was deferred?
-                    assert property_type is not None
-                    typ = property_type
-                else:
-                    typ = self.function_type(defn)
+                typ: Type = self.function_type(defn)
                 override_class_or_static = defn.is_class or defn.is_static
                 override_class = defn.is_class
             else:
                 assert defn.var.is_ready
                 assert defn.var.type is not None
                 typ = defn.var.type
-                if defn.func.is_property:
-                    assert isinstance(typ, ProperType) and isinstance(typ, CallableType)
-                    typ = typ.ret_type
                 override_class_or_static = defn.func.is_class or defn.func.is_static
                 override_class = defn.func.is_class
             typ = get_proper_type(typ)
@@ -1745,15 +1737,8 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 typ = bind_self(typ, self.scope.active_self_type(), is_classmethod=override_class)
             # Map the overridden method type to subtype context so that
             # it can be checked for compatibility.
-            if isinstance(base_attr.node, OverloadedFuncDef) and base_attr.node.is_property:
-                original_type = get_settable_property_type(base_attr.node)
-            else:
-                original_type = base_attr.type
-            original_type = get_proper_type(original_type)
+            original_type = get_proper_type(base_attr.type)
             original_node = base_attr.node
-            if isinstance(original_node, Decorator) and original_node.func.is_property:
-                if isinstance(original_type, CallableType):
-                    original_type = get_proper_type(original_type.ret_type)
             # `original_type` can be partial if (e.g.) it is originally an
             # instance variable from an `__init__` block that becomes deferred.
             if original_type is None or isinstance(original_type, PartialType):
@@ -1786,15 +1771,33 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 original_class_or_static = fdef.is_class or fdef.is_static
             else:
                 original_class_or_static = False  # a variable can't be class or static
+
+            if isinstance(original_type, FunctionLike):
+                original_type = self.bind_and_map_method(base_attr, original_type, defn.info, base)
+                if is_property(original_node):
+                    original_type = get_property_type(original_type)
+
+            if isinstance(typ, FunctionLike) and is_property(defn):
+                typ = get_property_type(typ)
+                if (
+                    isinstance(original_node, Var)
+                    and not original_node.is_final
+                    and (not original_node.is_property or original_node.is_settable_property)
+                    and isinstance(defn, Decorator)
+                ):
+                    # We only give an error where no other similar errors will be given.
+                    self.msg.fail(
+                        "Cannot override writeable attribute with read-only property", defn
+                    )
+
             if isinstance(original_type, AnyType) or isinstance(typ, AnyType):
                 pass
             elif isinstance(original_type, FunctionLike) and isinstance(typ, FunctionLike):
-                original = self.bind_and_map_method(base_attr, original_type, defn.info, base)
                 # Check that the types are compatible.
                 # TODO overloaded signatures
                 self.check_override(
                     typ,
-                    original,
+                    original_type,
                     defn.name,
                     name,
                     base.name,
@@ -1810,7 +1813,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 pass
             elif (
                 base_attr.node
-                and not self.is_writable_attribute(base_attr.node)
+                and not self.is_writable_attribute(original_node)
                 and is_subtype(typ, original_type)
             ):
                 # If the attribute is read-only, allow covariance
@@ -6084,6 +6087,8 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
     def is_writable_attribute(self, node: Node) -> bool:
         """Check if an attribute is writable"""
         if isinstance(node, Var):
+            if node.is_property and not node.is_settable_property:
+                return False
             return True
         elif isinstance(node, OverloadedFuncDef) and node.is_property:
             first_item = cast(Decorator, node.items[0])
@@ -6991,13 +6996,21 @@ def is_static(func: FuncBase | Decorator) -> bool:
     assert False, f"Unexpected func type: {type(func)}"
 
 
-def get_settable_property_type(defn: OverloadedFuncDef) -> Type | None:
-    assert isinstance(defn.items[0], Decorator)
-    first_type = defn.items[0].var.type
-    if first_type is None:
-        return None
-    assert isinstance(first_type, ProperType) and isinstance(first_type, CallableType)
-    return first_type.ret_type
+def is_property(defn: SymbolNode) -> bool:
+    if isinstance(defn, Decorator):
+        return defn.func.is_property
+    if isinstance(defn, OverloadedFuncDef):
+        if defn.items and isinstance(defn.items[0], Decorator):
+            return defn.items[0].func.is_property
+    return False
+
+
+def get_property_type(t: ProperType) -> ProperType:
+    if isinstance(t, CallableType):
+        return get_proper_type(t.ret_type)
+    if isinstance(t, Overloaded):
+        return get_proper_type(t.items[0].ret_type)
+    return t
 
 
 def is_subtype_no_promote(left: Type, right: Type) -> bool:

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -1720,8 +1720,17 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 context = defn.func
 
             # Construct the type of the overriding method.
+            # TODO: this logic is much less complete than similar one in checkmember.py
             if isinstance(defn, (FuncDef, OverloadedFuncDef)):
-                typ: Type = self.function_type(defn)
+                if isinstance(defn, OverloadedFuncDef) and defn.is_property:
+                    # Type for a settable property is stored in a non-trivial way.
+                    first = defn.items[0]
+                    assert isinstance(first, Decorator)
+                    first_type = first.var.type
+                    assert isinstance(first_type, ProperType) and isinstance(first_type, CallableType)
+                    typ: Type = first_type.ret_type
+                else:
+                    typ = self.function_type(defn)
                 override_class_or_static = defn.is_class or defn.is_static
                 override_class = defn.is_class
             else:
@@ -4311,7 +4320,8 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 if len([k for k in sig.arg_kinds if k.is_required()]) > 1:
                     self.msg.fail("Too many arguments for property", e)
             self.check_incompatible_property_override(e)
-        if e.func.info and not e.func.is_dynamic():
+        # For overloaded functions we already checked override for overload as a whole.
+        if e.func.info and not e.func.is_dynamic() and not e.is_overload:
             self.check_method_override(e)
 
         if e.func.info and e.func.name in ("__init__", "__new__"):

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -1774,7 +1774,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
             if isinstance(original_type, FunctionLike):
                 original_type = self.bind_and_map_method(base_attr, original_type, defn.info, base)
-                if is_property(original_node):
+                if original_node and is_property(original_node):
                     original_type = get_property_type(original_type)
 
             if isinstance(typ, FunctionLike) and is_property(defn):
@@ -1812,7 +1812,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 #
                 pass
             elif (
-                base_attr.node
+                original_node
                 and not self.is_writable_attribute(original_node)
                 and is_subtype(typ, original_type)
             ):

--- a/test-data/unit/check-abstract.test
+++ b/test-data/unit/check-abstract.test
@@ -789,7 +789,7 @@ class A(metaclass=ABCMeta):
     def x(self) -> int: pass
 class B(A):
     @property
-    def x(self) -> str: pass # E: Return type "str" of "x" incompatible with return type "int" in supertype "A"
+    def x(self) -> str: pass # E: Signature of "x" incompatible with supertype "A"
 b = B()
 b.x() # E: "str" not callable
 [builtins fixtures/property.pyi]

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -7366,3 +7366,26 @@ class D(C[List[T]]): ...
 di: D[int]
 reveal_type(di)  # N: Revealed type is "Tuple[builtins.list[builtins.int], builtins.list[builtins.int], fallback=__main__.D[builtins.int]]"
 [builtins fixtures/tuple.pyi]
+
+[case testOverrideAttrWithSettableProperty]
+class Foo:
+    def __init__(self) -> None:
+        self.x = 42
+
+class Bar(Foo):
+    @property
+    def x(self) -> int: ...
+    @x.setter
+    def x(self, value: int) -> None: ...
+[builtins fixtures/property.pyi]
+
+[case testOverrideAttrWithSettablePropertyAnnotation]
+class Foo:
+    x: int
+
+class Bar(Foo):
+    @property
+    def x(self) -> int: ...
+    @x.setter
+    def x(self, value: int) -> None: ...
+[builtins fixtures/property.pyi]

--- a/test-data/unit/check-dataclasses.test
+++ b/test-data/unit/check-dataclasses.test
@@ -1286,7 +1286,7 @@ class A:
 @dataclass
 class B(A):
     @property
-    def foo(self) -> int: pass  # E: Signature of "foo" incompatible with supertype "A"
+    def foo(self) -> int: pass
 
 reveal_type(B)  # N: Revealed type is "def (foo: builtins.int) -> __main__.B"
 

--- a/test-data/unit/check-inference.test
+++ b/test-data/unit/check-inference.test
@@ -1475,9 +1475,8 @@ class A:
         self.x = [] # E: Need type annotation for "x" (hint: "x: List[<type>] = ...")
 
 class B(A):
-    # TODO?: This error is kind of a false positive, unfortunately
     @property
-    def x(self) -> List[int]:  # E: Signature of "x" incompatible with supertype "A"
+    def x(self) -> List[int]:
         return [123]
 [builtins fixtures/list.pyi]
 

--- a/test-data/unit/check-inference.test
+++ b/test-data/unit/check-inference.test
@@ -1475,7 +1475,7 @@ class A:
         self.x = [] # E: Need type annotation for "x" (hint: "x: List[<type>] = ...")
 
 class B(A):
-    @property
+    @property  # E: Cannot override writeable attribute with read-only property
     def x(self) -> List[int]:
         return [123]
 [builtins fixtures/list.pyi]

--- a/test-data/unit/check-inference.test
+++ b/test-data/unit/check-inference.test
@@ -1475,8 +1475,8 @@ class A:
         self.x = [] # E: Need type annotation for "x" (hint: "x: List[<type>] = ...")
 
 class B(A):
-    @property  # E: Cannot override writeable attribute with read-only property
-    def x(self) -> List[int]:
+    @property
+    def x(self) -> List[int]:  # E: Cannot override writeable attribute with read-only property
         return [123]
 [builtins fixtures/list.pyi]
 

